internal/services/serviceprincipals/service_principal_password_resource_msgraph.go

```diff
@@ -0,0 +1,160 @@
+				return tf.ImportAsExistsDiag("azuread_service_principal_password", id.String())
+			}
+			newCredentials = append(newCredentials, cred)
+		}
+	}
+
+	newCredentials = append(newCredentials, *credential)
+
+	properties := msgraph.ServicePrincipal{
+		ID:                  &id.ObjectId,
+		PasswordCredentials: &newCredentials,
+	}
+	if _, err := client.Update(ctx, properties); err != nil {
+		return tf.ErrorDiagF(err, "Adding password for service principal with object ID %q", id.ObjectId)
+	}
+
+	d.SetId(id.String())
+
+	return servicePrincipalPasswordResourceReadMsGraph(ctx, d, meta)
+}
+
+func servicePrincipalPasswordResourceReadMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics { //nolint
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+
+	id, err := parse.PasswordID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing password credential with ID %q", d.Id())
+	}
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "service_principal_id", "Service principal with object ID %q was not found", id.ObjectId)
+		}
+		return tf.ErrorDiagPathF(err, "service_principal_id", "Retrieving service principal with object ID %q", id.ObjectId)
+	}
+
+	var credential *msgraph.PasswordCredential
+	if app.PasswordCredentials != nil {
+		for _, cred := range *app.PasswordCredentials {
+			if cred.KeyId != nil && *cred.KeyId == id.KeyId {
+				credential = &cred
+				break
+			}
+		}
+	}
+
+	if credential == nil {
+		log.Printf("[DEBUG] Password credential %q (ID %q) was not found - removing from state!", id.KeyId, id.ObjectId)
+		d.SetId("")
+		return nil
+	}
+
+	tf.Set(d, "description", credential.DisplayName)
+	tf.Set(d, "display_name", credential.DisplayName)
+	tf.Set(d, "key_id", id.KeyId)
+	tf.Set(d, "service_principal_id", id.ObjectId)
+
+	startDate := ""
+	if v := credential.StartDateTime; v != nil {
+		startDate = v.Format(time.RFC3339)
+	}
+	tf.Set(d, "start_date", startDate)
+
+	endDate := ""
+	if v := credential.EndDateTime; v != nil {
+		endDate = v.Format(time.RFC3339)
+	}
+	tf.Set(d, "end_date", endDate)
+
+	return nil
+}
+
+func servicePrincipalPasswordResourceDeleteMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics { //nolint
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+
+	id, err := parse.PasswordID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing password credential with ID %q", d.Id())
+	}
+
+	tf.LockByName(servicePrincipalResourceName, id.ObjectId)
+	defer tf.UnlockByName(servicePrincipalResourceName, id.ObjectId)
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			log.Printf("[DEBUG] Service Principal with Object ID %q was not found - removing from state!", id.ObjectId)
+			return nil
+		}
```

Comment:
(as above) this is hiding a bug which should be in the refresh/read - can we remove this